### ISSUE TITLE
Add 429 retry capability

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#9550]
**What would you like Renovate to be able to do?**

Retry - within reason - when receiving a 429 response from remote servers.

**Did you already have any implementation ideas?**

It should be configurable via `hostRules` and probably disabled by default (i.e. current behavior).

There's two major things to think about:
1. The retry-after interval may be too high (e.g. an hour) and we don't want to retry no matter what
2. Even if the interval is OK, we don't want to retry forever

Keeping state for this may be difficult, but is ideal.

Maybe two fields:
- defaultRetryAfter, e.g. 5,000ms. Only to be used if no retry-interval returned
- maximumRetryTotal, e.g. 60,000ms

If no retry-after value is received, we use defaultRetryAfter. Then, we check if retryAfter < maximumRetryTotal. If it is, then we wait before retrying again and reduce the remaining retry total accordingly. If it's not, then we return 429 to the requester.

**Comment by rarkins** [comment]
Will move from requirements to ready once others have commented and confirmed this approach, and if we should do it by default or not.

**Comment by HonkingGoose** [comment]
Did you consider the "exponential backoff algorithm" do deal with the `429 Too many requests` response?

> An exponential backoff algorithm retries requests exponentially, increasing the waiting time between retries up to a maximum backoff time. For example:
>  1. Make a request to Cloud IoT Core.
>  2. If the request fails, wait 1 + random_number_milliseconds seconds and retry the request.
>  3. If the request fails, wait 2 + random_number_milliseconds seconds and retry the request.
>  4. If the request fails, wait 4 + random_number_milliseconds seconds and retry the request.
>  5. And so on, up to a maximum_backoff time.
> 6. Continue waiting and retrying up to some maximum number of retries, but do not increase the wait period between retries.

Source for quote: https://cloud.google.com/iot/docs/how-tos/exponential-backoff
The links page also has Node.js examples.

I don't know if this is something you like/hate. Just throwing this out there for consideration, as it seems relevant and quite elegant.

**Comment by rarkins** [comment]
If the remote end tells us how much time to wait then we should just use that. When there is no retry time specified.. we could consider the exponential back-off. In the bitbucket example in #9441 the jenkins plugin uses a constant 5s retry delay.

**Comment by fgreinacher** [comment]
+1 for exponential backoff as suggested by @HonkingGoose. Constant retry intervals sometimes make things worse because they don't give the host a chance to recover/scale/... If the `Retry-After` header is set Renovate should of course honor it.

**Comment by froque** [comment]
Jenkins plugin is also considering an exponential backoff: https://issues.jenkins.io/browse/JENKINS-64418

**Comment by HonkingGoose** [comment]
Copy/pasting a relevant post from #10133 by @AMDmi3 (Repology server maintainer):

> 
> 
> I don't think I set `retry-after`, and not sure if it can be done meaningfully.
> I can imagine it as follows:
> 
>     * say I'm ok with 1RPS so I set `retry-after: 1`
> 
>     * client does 20 parallel requests
> 
>     * gets one 200 reply and and 19 429s with `retry-after: 1`
> 
>     * client waits a second and does 19 parallel requests...
> 
> 
> IMO the correct behavior for the client would be to just decrease the rate multiplicatively on 429s and remember that rate to not hit the limit again.
> Another approach would be to use some pessimistic rate by default and set higher rate manually for individual hosts which require and allow it. I do that in repology-linkchecker and am not hitting any 429s at all, though it does millions requests to thousands of hosts.

**Comment by AMDmi3** [comment]
A small additional thought - if you disable parallel requests on 429 right away, the behavior suddenly becomes predictable and following `retry-after` works correctly. So alternate approach would be to disable parallel requests on first 429, then increase wait time according to `retry-after` (or some meaningful multiplier like 1.25x or 1.5x if it's not provided) on following 429s, if any.

**Comment by okainov** [comment]
FYI it's not only `429`, but also `403` (in case of Github, for example)

```
16:30:43               "response": {
16:30:43                 "statusCode": 403,
16:30:43                 "statusMessage": "Forbidden",
16:30:43                 "body": {
16:30:43                   "documentation_url": "https://docs.github.com/en/free-pro-team@latest/rest/overview/resources-in-the-rest-api#secondary-rate-limits",
16:30:43                   "message": "You have exceeded a secondary rate limit. Please wait a few minutes before you try again."
16:30:43                 },
16:30:43                 "headers": {
16:30:43                   "server": "GitHub.com",
16:30:43                   "date": "Mon, 27 Sep 2021 14:30:32 GMT",
16:30:43                   "content-type": "application/json; charset=utf-8",
16:30:43                   "transfer-encoding": "chunked",
16:30:43                   "retry-after": "60",
16:30:43                   "access-control-expose-headers": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset",
16:30:43                   "access-control-allow-origin": "*",
16:30:43                   "x-github-media-type": "github.v3; format=json",
16:30:43                   "strict-transport-security": "max-age=31536000; includeSubdomains; preload",
16:30:43                   "x-frame-options": "deny",
16:30:43                   "x-content-type-options": "nosniff",
16:30:43                   "x-xss-protection": "0",
16:30:43                   "referrer-policy": "origin-when-cross-origin, strict-origin-when-cross-origin",
....
```

**Comment by rarkins** [comment]
We should double check whether the `got` library we use happens to be doing it already or not (ping @szmarczak can you confirm?). I wrote earlier that our default behavior is *not* to retry, but got some doubts.

If we *don't* already retry in any case, implementation of this feature needs the following considerations:
- If done over simply, it would result in *more* requests being sent than before, to endpoints which are already complaining to us we are sending too fast. i.e. we'd retry - according to their limits - when today we instead give up
- To do it well, we'd need to keep state per-host. This is maybe even something we should defer to `got` for instead. "Giving up" a single request on a 4xx instead of retrying won't be that helpful if we have another 20 requests queued up which we proceed to send anyway.

Seen more holistically, I think we need:
- per-host state to remember if it's in a state where it told us to stop retrying for some period (note: this is technically per-host per-token in some cases, but I'm not sure we have any scenarios where we're hitting limits and have different tokens in use)
- option of whether to retry to give up when receiving a 4xx with retry-after
- option to set maximum retry-after. e.g. if github returns retry-after 3600s you probably don't want to sit there idle for an hour doing no work 
- option to limit requests per window of time (e.g. 0.5 requests per second, meaning 2s gap per request) to hopefully avoid the rate limit in the first place [and we should ship with defaults for certain public, well-known hosts like GitHub or Repology)

In short: still needs some research/discussion. Anyone not interested should Unsubscribe :)

**Comment by rarkins** [comment]
@szmarczak thanks.

Current docs: https://github.com/sindresorhus/got/blob/215e06a4993329578d92d4f44607913239a03094/documentation/7-retry.md

- defaults to two retries
- includes 429 in the list of codes

Do I understand it correctly that it's enabled by default, i.e. in use by Renovate today?

**Comment by HonkingGoose** [comment]
@rarkins Maybe we can remove the preset `workarounds:reduceRepologyServerLoad` that was added as a "stopgap" measure with PR #10146?

**Comment by rarkins** [comment]
No, because it helps to reduce the number of 429's in the first place
